lib/livebook/apps/manager.ex

```diff
@@ -0,0 +1,245 @@
+        case fetch_app(app_spec.slug) do
+          {:ok, _state, app} when app.app_spec.version == app_spec.version ->
+            state
+
+          {:ok, :reachable, app} ->
+            ref = redeploy(app, app_spec)
+            track_deployment(state, app_spec, ref)
+
+          {:ok, :unreachable, _app} ->
+            state
+
+          :error ->
+            ref = deploy(app_spec)
+            track_deployment(state, app_spec, ref)
+        end
+    end
+  end
+
+  defp close_leftover_apps(permanent_app_specs) do
+    permanent_slugs = MapSet.new(permanent_app_specs, & &1.slug)
+
+    for app <- Apps.list_apps(),
+        app.permanent,
+        app.slug not in permanent_slugs do
+      Livebook.App.close_async(app.pid)
+    end
+  end
+
+  defp fetch_app(slug) do
+    # We check both global and the tracker. The app may be present in
+    # the tracker for longer, but if it is actually down we will get
+    # the :app_closed event eventually. On the other hand, if the app
+    # is already somewhere, global will tell us sooner than tracker.
+    case Apps.fetch_app(slug) do
+      {:ok, app} ->
+        {:ok, :reachable, app}
+
+      :error ->
+        case Livebook.Tracker.fetch_app(slug) do
+          {:ok, app} -> {:ok, :unreachable, app}
+          :error -> :error
+        end
+    end
+  end
+
+  defp app_definitely_down?(slug) do
+    not Apps.exists?(slug) and Livebook.Tracker.fetch_app(slug) == :error
+  end
+
+  defp deploy(app_spec) do
+    deployer_pid = Livebook.Apps.Deployer.list_deployers() |> Enum.random()
+    Logger.info("[app=#{app_spec.slug}] Scheduling app deployment on node #{node(deployer_pid)}")
+
+    Livebook.Apps.Deployer.deploy_monitor(deployer_pid, app_spec,
+      permanent: true,
+      start_only: true
+    )
+  end
+
+  defp redeploy(app, app_spec) do
+    # Redeploying pushes the new notebook to an existing app process,
+    # so we need to run deployment on the same node
+    node = node(app.pid)
+    deployer_pid = Livebook.Apps.Deployer.local_deployer(node)
+    Logger.info("[app=#{app_spec.slug}] Scheduling app deployment on node #{node(deployer_pid)}")
+    Livebook.Apps.Deployer.deploy_monitor(deployer_pid, app_spec, permanent: true)
+  end
+
+  defp track_deployment(state, app_spec, ref) do
+    put_in(state.deployments[app_spec.slug], %{
+      ref: ref,
+      retries: 0,
+      app_spec: app_spec,
+      slug: app_spec.slug
+    })
+  end
+
+  defp handle_deployment_failure(state, deployment, message) do
+    Logger.error("[app=#{deployment.slug}] Deployment failed, #{message}")
+
+    # Schedule retry
+    %{app_spec: app_spec, retries: retries} = deployment
+    retries = retries + 1
+    time = @retry_backoff_base_ms * min(retries, 6)
+    Process.send_after(self(), {:retry, app_spec.slug}, time)
+    put_in(state.deployments[app_spec.slug].retries, retries)
+  end
+
+  defp retry(state, slug) do
+    if app_definitely_down?(slug) do
+      %{app_spec: app_spec} = state.deployments[slug]
```

Comment:
If you decide to retry, and then the app is undeployed, it will no longer be here. We should probably deal with such cases?